### ISSUE TITLE
[BUG] KafkaConsumer.position() Method should give the offset one larger than the last record

### ISSUE THREAD

**Issue by WalterMa** [streamnative/kop#276]
**Describe the bug**
According to Kafka Conusmer API Doc:
https://kafka.apache.org/22/javadoc/index.html?org/apache/kafka/clients/consumer/KafkaConsumer.html

> The position of the consumer gives the offset of the next record that will be given out. It will be **one larger than the highest offset the consumer has seen** in that partition. It automatically advances every time the consumer receives messages in a call to poll(Duration).

While currently, KoP will return the offset of the last record, without adding one to it

**To Reproduce**
1. Suscribe to a Kafka topic or a KoP topic
2. Call `conusmer.seekToEnd()` and then print `conusmer.position()` result
3. Kafka will return the offset one larger than last record, while KoP not.

**Expected behavior**
KafkaConsumer.position() Method should give the offset one larger than the last record

**Comment by BewareMyPower** [comment]
Could you give your code example? I accidentally found KoP doesn't support `KafkaConsumer#seekEnd` yet. The error is

> java.lang.IllegalStateException: You can only check the position for partitions assigned to this consumer.

**Comment by BewareMyPower** [comment]
Sorry it's my fault. I've set the wrong partition in my tests before. I'll take a look at this issue.

**Comment by WalterMa** [comment]
> Sorry it's my fault. I've set the wrong partition in my tests before. I'll take a look at this issue.

That's great!
Do you still need some code examples for reproducing?

**Comment by BewareMyPower** [comment]
@WalterMa no, I've reproduced the problem in a unit test. Also I found there're no tests for `KafkaConsumer#seek`. If the bug was fixed, the related tests would be added as well.

**Comment by BewareMyPower** [comment]
I've a question. What does the **wrong** behavior affects?

Currently the KoP's offset is just mapped to Pulsar's **message id**, i.e. the `(ledger id, entry id, batch index)` triple. So it makes offset not continuous.

It's different from Kafka which uses a HW(High Watermark, **the next uncommitted offset**) to fetch messages. KoP just fetches messages from the latest message id, i.e., **the latest committed offset**. It's because KoP or Pulsar doesn't know what the next message's id is. But Kafka knows the next message's offset is just `lastOffset + 1`.

In conclusion, if there's not a big issue, we should change the current behavior after continuous offset is implemented.

**Comment by WalterMa** [comment]
I found this wrong behavior, when I was trying to adapt spark-streaming-kafka-010 to KoP ( to solve repeated conume in the NonSectiveMode).

Spark Kafka Direct Stream will poll messages in **[fromOffset, untilOffset)**
The fromOffset value = previous untilOffset
untilOffset = conumser.position()

Let's look at an example, assume lastMsgOffset = 1000:
In Kafka, [fromOffset, untilOffset) = [ anyVal<1000, **1001**)
In KoP, [fromOffset, untilOffset) = [ anyVal<1000, **1000**)

When next messages comes, 
In Kafka, Spark will `seek to 1001, then poll()`, and **get next messages**.
In KoP, Spark will `seek to 1000, then poll`, and **get the previous lastest message and next messages**.  (since we start poll from offset=1000, then lead to repeated conume)

**Comment by WalterMa** [comment]
As for how to fix this issure, please allow me to make a few comments. I'am not familiar with this project, so if any wrong, please correct me.

Can we add one to lastOffset only for the postsion() request ? 

> when consumer send FETCH requests, KoP did like fetchFromOffset(request.offset - 1)     

, and I'am a little comfused why we need to ` request.offset - 1` when fetch.

**Comment by BewareMyPower** [comment]
Got it. I'll try to fix this issue after KoP 2.7.0 is released.

----

> and I'am a little comfused why we need to request.offset - 1 when fetch.

When Kafka handles fetch requests, it reads data **from** `request.offset`.

However when KoP handles fetch requests, the workflow is:

1. Convert `request.offset` to a message id;
2. Read data **after** the message id.

Kafka consumer maintains an internal offset to construct FETCH request. The offset will be updated after `resetOffsetsAsync`. Currently KoP returns a `latestOffset` not `latestOffset + 1` if consumer resets offset to latest. If we change it to `latestOffset + 1`, the consumer may fetch from `latestOffset + 1`.
